# Scarus_rivulatus

DNA metabarcoding is increasingly used in dietary studies, but this molecular technique has limitations including DNA extraction, primer choice suitability and completeness of DNA reference databases. Together, these limitations can create successive biases, some of which can be identified by careful data interrogation. Here, we examine some of the influential factors impacting DNA metabarcoding by focusing on the diet of the grazing coral reef parrotfish Scarus rivulatus (Labridae). The 16S V3-4 and 18S V4 gene regions were targeted to detect a wide range of taxa in the pharyngeal content, but genus-level amplicon sequence variants (ASVs) assignments were â‰¤ 40%. Multiple universal 18S V4 primer sets, which amplified the same portion of the gene region, highlighted disparities that impacted resulting datasets and dietary data interpretation. Approximately 50% of ASVs overlapped among these 18S V4 datasets. Overall, our results show that protein-rich epilithic and endolithic microscopic photoautotrophs were consistently present and that S. rivulatus appears to rely on protein-rich taxa, especially dinoflagellates and cyanobacteria. Even if all qualitative limitations could be addressed, our data emphasise the difficulty of making straightforward quantitative conclusions about the diet composition of grazing fishes using DNA metabarcoding alone. Thus, until DNA metabarcoding reaches a state of maturity where quantitative data can be gained, additional support is required from other dietary analysis methods to provide robust interpretations and conclusions. 
